# Исследование надёжности заёмщиков

Описание проекта\
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.\
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Описание данных

- children — количество детей в семье

- days_employed — общий трудовой стаж в днях

- dob_years — возраст клиента в годах

- education — уровень образования клиента

- education_id — идентификатор уровня образования

- family_status — семейное положение

- family_status_id — идентификатор семейного положения

- gender — пол клиента

- income_type — тип занятости

- debt — имел ли задолженность по возврату кредитов

- total_income — ежемесячный доход

- purpose — цель получения кредита

In [42]:
import pandas as pd

try:
    data = pd.read_csv('data_borrowers.csv')
except:
    print('error')

In [43]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [44]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [45]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце я обработаю на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` будет заполнен медианным доходом среди всех записей с тем же типом.**

In [46]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений


**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`.Обработаю значения в этом столбце: заменю все отрицательные значения положительными с помощью метода `abs()`.**

In [47]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведу медианное значение трудового стажа `days_employed` в днях.**

In [48]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Перечень уникальных значений столбца `children`.**

In [49]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [50]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [51]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)


**Заполню пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [52]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [53]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [54]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обработаю неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведу их к нижнему регистру и проверю остальные столбцы.**

In [55]:
data['education'] = data['education'].str.lower()

In [56]:
data.duplicated().sum()

71

In [57]:
data = data.drop_duplicates()

### Категоризация данных


** На основании диапазонов, указанных ниже, создам в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [58]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [59]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [60]:
data['purpose'].unique() # уникальные цели взятия кредита

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создам функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [61]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [62]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследование данные 

#### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [63]:
print(data.isna().sum()) # проверим налиичие пропусков 


children                 0
days_employed            0
dob_years                0
education                0
education_id             0
family_status            0
family_status_id         0
gender                   0
income_type              0
debt                     0
total_income             0
purpose                  0
total_income_category    0
purpose_category         0
dtype: int64


In [64]:
print(data.sort_values(by = 'children')['children'].unique())

[0 1 2 3 4 5]


Аномальных или отрицательных значений нет, значит ничего не изменяем

In [65]:
d1 = data.groupby('children')['debt'].sum() #количество клиентов с задолженносьбю с определенным количеством детей
d1

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [66]:
d2 =data.groupby('children')['debt'].count() # всего клиентов в данных с определенным количеством детей
d2

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

Посчитаем конверсию задолженности для каждой категории клиентов по количеству людей

In [67]:
d3 = d1/d2
columns = (f"{d3[i]:.2%}" for i in range(len(d3)))
df1 = pd.Series(columns)
df1

0    7.54%
1    9.23%
2    9.45%
3    8.18%
4    9.76%
5    0.00%
dtype: object

__Вывод: Как можно заметить конверсия о задолженности кредита для клиентов с определенным количеством детей от 1 до 4 почти одинакова, a минимальная конверсия у килиентов, которые не имеют детей(7.54%), максимальная у клиентов, у которых 4 ребенка(9.76%). У клиентов с пятью детьми конверсия равна нулю, но из-за этого нельзя сказать, что все клиенты с пятью детьми будут возвращать кредит всегда в срок, так как количестов клиентов с пятью детьми всего 9 в базе данных, что не сравнится, например, с количеством клиентов у кого 2 ребенка, то есть 2052. Чёткой зависимости найти не удалось.__

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [68]:
print(data.sort_values('family_status')['family_status'].unique())

['Не женат / не замужем' 'в разводе' 'вдовец / вдова' 'гражданский брак'
 'женат / замужем']


In [69]:
# проверим на дубликаты 
data['family_status'] = data['family_status'].str.lower()
data.duplicated().sum()

0

In [70]:
data.sort_values('family_status_id')['family_status_id'].unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [71]:
# family_status_id - категориальные данные для family_status
print(data['family_status'].value_counts())
print(data['family_status_id'].value_counts())

женат / замужем          12261
гражданский брак          4134
не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64
0    12261
1     4134
4     2796
3     1189
2      951
Name: family_status_id, dtype: int64


По количетву сопоставим id и название категории <br>
женат / замужем - 0<br>
гражданский брак - 1<br>
вдовец / вдова - 2<br>
в разводе - 3<br>
не женат / не замужем - 4<br>


In [72]:
e1 = data.groupby('family_status_id')['debt'].sum() # количестов клиентов по категориям с заолженностью
e1

family_status_id
0    927
1    385
2     63
3     84
4    273
Name: debt, dtype: int64

In [73]:
e2 = data.groupby('family_status_id')['debt'].count() # всего количестов клиентов по категориям
e2

family_status_id
0    12261
1     4134
2      951
3     1189
4     2796
Name: debt, dtype: int64

Посчитаем конверсию для каждой категории клиентов

In [74]:
e3 = e1/e2
columns1 = list(f"{e3[i]:.2%}" for i in range(len(e3)))
columns2 = ['женат / замужем','гражданский брак','вдовец / вдова','в разводе','не женат / не замужем']
df2 = pd.DataFrame({'percents': columns1, 'status': columns2})
df2

percents                 status
0    7.56%        женат / замужем
1    9.31%       гражданский брак
2    6.62%         вдовец / вдова
3    7.06%              в разводе
4    9.76%  не женат / не замужем

__Вывод: Как можно заметить у клиентов, которые принадлежат к статусу вдовец/вдова, проблем с возвращением кредита в срок меньше всего(6.62%), а у клиентов, с категорией не женат/не замужем, процент не возвращения кредита в срок самый высокий(9.76%). Объеденяя в группы клиентов по наличию партнеров, нельзя сказать, что между ними наблюдается чёткая зависимость.__ 

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [75]:
print(len(data['total_income'].unique()))  

18504


Количестов различных значений дохода большое, значит надо категоризировать данные

In [76]:
data.sort_values(by = 'total_income')['total_income']

14585      20667
13006      21205
16174      21367
1598       21695
14276      21895
          ...   
17178    1711309
20809    1715018
9169     1726276
19606    2200852
12412    2265604
Name: total_income, Length: 21331, dtype: int32

Общий доход варируется от 20667 до 2265604, отрицательных значений нет, аномальных тоже

In [77]:
r3 = data.groupby('total_income_category')['debt'].mean()

Снова посчитаем конверсию для кааждой категории клиентов по задолженности кредита

In [78]:
print(r3)
columns = (f"{r3[i]:.2%}" for i in range(len(r3)))
df3 = pd.Series(columns)
df3

total_income_category
A    0.080000
B    0.070602
C    0.084982
D    0.060172
E    0.090909
Name: debt, dtype: float64


0    8.00%
1    7.06%
2    8.50%
3    6.02%
4    9.09%
dtype: object

**Вывод: Самый низкий показатель по задолженности по кредиту у клиентов категории доходов D и B. Самые высокие показатели у клиентов A, C, E категорий. Таким образом самыми "безопасными" клиентами для банка являются люди, чей доход составляет от 30 до 50 тысяч, самыми "опасными" клиенты с доходом от 0 до 30 тысяч.** 

#### Как разные цели кредита влияют на его возврат в срок?

In [79]:
t1 = data.groupby('purpose_category')['debt'].sum() # всего клиентов по категориям 
t1

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

In [80]:
t2 = data.groupby('purpose_category')['debt'].count() # количество клиентов по категориям, которые имели задолжность
t2

purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64

Найдем конверсию

In [81]:
t3 = t1/t2
print(t3)
columns = (f"{t3[i]:.2%}" for i in range(len(t3)))
df4 = pd.Series(columns)
df4

purpose_category
операции с автомобилем      0.093480
операции с недвижимостью    0.072551
получение образования       0.092528
проведение свадьбы          0.079118
Name: debt, dtype: float64


0    9.35%
1    7.26%
2    9.25%
3    7.91%
dtype: object

**Вывод: Самые низкие показатели по задолженности по кредиту у клиентов, чьи цели были связаны с опрациями с недвижимостью и проведением свадьбы. Самыми высокими показателями у клиентов, чьи цели были связаны с получением образования и операциями с автомобилем.** 

#### Возможные причины появления пропусков в исходных данных.


*В данных пропущены значения трудового стажа в днях и доходе, причем количество пропусков совпадает и является не таким большим от общего количества данных. Мне кажется, что эти пропуски не являются техниической причиной, а сделаны из-за человеческого фактора, так как некоторые люди могут не помнить количество рабочих дней, а некотоыре могут не хотеть раскрывать сумму своего дохода.* 

#### Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Медианное значение - лучшее решение для заполнения пропусков количественных переменных, так как медиана - это значение выборки ровно по середине, где остальное количество значний меньших медианы равно количеству остальных значений, которые больше медианы. Также на медиану не влияют аномальные значения, в отличие от среднего арифметического.* 

### Общий вывод.

__Общий вывод:__<br>
Для решения поставленной задачи было выведено и проверено несколько гипотез на основании входных данных.

* Количество детей влияет на появление задолженности по кредиту.
* Семейное положение клиента влияет на появление задолженности по кредиту.
* Уровень дохода влияет на появление задолженности по кредиту.
* цели взятия кредита влияют на появление заодолженности по нему.



__Гипотеза 1.__ <br>
В данных действительно найдена закономерность: клиенты, у которых нет детей имеют самый маленький процент задолженности по кредиту, что составляет 7.54%. Клиенты, которые имеют от 1 до 4 детей больше подвержены задержанию выплаты кредита. Однако классы в выборке имеют разные размеры, то есть выборка не сбалансированна, а разница значений конверсии для разных категорий не слишком велика.
* 1 ребенок - 9.23%
* 2 ребенка - 9.45%
* 3 ребенка - 8.18%
* 4 ребенка - 9.76%
* Для клиентов с 5 детьмя вывод о возврате кредита сделать нельзя, так как данных слишком мало. <br>


__Гипотеза 2.__ <br>
Категоризируя клиентов по наличию\отсутствия партнеров нельзя сказать, что есть четкая зависимость по задолженности, однако, клиенты, которые не состоят или не состояли в браке, являются "опасными" для банка. Получается, отношения с зарегистрированным браком, положительно влияют на выплату кредита в срок.
<br>
* 7.56%	женат / замужем
* 9.31%	гражданский брак
* 6.62%	вдовец / вдова
* 7.06%	в разводе
* 9.76%	не женат / не замужем<br>


__Гипотеза 3.__ <br>
Гипотеза частично подтверждена,однако для более чёткой зависимости требуется большее количестов данных. Самые безопасыне для банка клиенты категории доходов D и B, а самые опасные клиенты с категориями A, C, E. Чёткой зависимости не найдено, требуется большего количества данных.
* E - 8.00% - доход < 30000
* D - 7.06% - доход от 30000 до 50000 
* C - 8.50% - доход от 50000 до 200000 
* B - 6.02% - доход от 200000 до 1000000
* A - 9.09% - доход >1000000<br>

__Гипотеза 4.__ <br>
Зависимость между этими катеогрияим обозначена. Самые низкие показатели по задолженности по кредиту у клиентов, чьи цели были связаны с опрациями с недвижимостью и проведением свадьбы. Самыми высокими показателями у клиентов, чьи цели были связаны с получением образования и операциями с автомобилем.
* операции с автомобилем - 9.35%
* операции с недвижимостью - 7.26%
* получение образования - 9.25%
* проведение свадьбы - 7.91%



